icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,418 @@ namespace icicle {
+
+          // update m_map_segment_id_2_inputs with the data that is ready for process
+          auto cur_segment_iter = m_map_segment_id_2_inputs.find(cur_segment_id);
+          cur_segment_iter->second->m_nof_inputs_ready +=
+            m_layers[completed_layer_idx].m_hash.output_size() * NOF_OPERATIONS_PER_TASK;
+
+          // check if cur segment is ready to be executed
+          const Hash cur_hash = m_layers[cur_layer_idx].m_hash;
+          const uint64_t nof_hashes_in_seg = std::min(
+            m_layers[cur_layer_idx].m_nof_hashes - cur_segment_idx * NOF_OPERATIONS_PER_TASK,
+            uint64_t(NOF_OPERATIONS_PER_TASK));
+          //          ICICLE_ASSERT(
+          //             > 0) << "Edge case negative number of hashes"; // Koren what is that for?
+          if (cur_segment_iter->second->m_nof_inputs_ready >= cur_hash.input_default_chunk_size() * nof_hashes_in_seg) {
+            const std::byte* task_input = (completed_layer_idx < m_output_store_min_layer)
+                                            ? cur_segment_iter->second->m_input_data
+                                            : &(m_layers[completed_layer_idx].m_results[0]);
+            dispatch_task(task, cur_layer_idx, cur_segment_idx, task_input);
+            continue;
+          }
+        }
+
+        if (l0_segment_idx < nof_segments_at_l0) {
+          // send a task from layer 0
+          dispatch_task(task, 0, l0_segment_idx, leaves);
+          l0_segment_idx++;
+          continue;
+        }
+        task->set_idle();
+      }
+      return eIcicleError::SUCCESS;
     }
 
-    std::pair<std::byte*, size_t> get_merkle_root() const override
+    std::pair<const std::byte*, size_t> get_merkle_root() const override
     {
-      ICICLE_LOG_DEBUG << "CPU CPUMerkleTreeBackend::get_merkle_root() called";
-      return {nullptr, 0}; // TODO: Implement root retrieval logic
+      ICICLE_ASSERT(m_tree_already_built) << "Merkle tree cannot provide root before built";
+      return {m_layers.back().m_results.data(), m_layers.back().m_hash.output_size()};
     }
 
     eIcicleError get_merkle_proof(
       const std::byte* leaves,
-      uint64_t element_idx,
+      uint64_t leaves_size,
+      uint64_t leaf_idx,
+      bool is_pruned,
       const MerkleTreeConfig& config,
       MerkleProof& merkle_proof) const override
     {
-      ICICLE_LOG_DEBUG << "CPU CPUMerkleTreeBackend::get_merkle_proof() called for element index " << element_idx;
-      // Dummy implementation. TODO implement
-      uint64_t root = 123;
-      merkle_proof.allocate(
-        false /*pruned*/, element_idx, leaves + element_idx * get_leaf_element_size(), get_leaf_element_size(),
-        (const std::byte*)&root, sizeof(root));
-      return eIcicleError::SUCCESS; // TODO: Implement proof generation logic
+      if (!m_tree_already_built) {
+        ICICLE_LOG_ERROR << "Tree cannot generate a merkle proof before built\n";
```

Comment:
again maybe this is a warning and not error

Reply:
same